src/main/java/Duke.java

```diff
@@ -1,10 +1,232 @@
+        horizontalLine();
+        boolean isLessThanOne = (tasks.size() <= 1);
+        System.out.println("Got it. I've removed this task:\n" + t.toString());
+        System.out.println("Now you have " + tasks.size() + (isLessThanOne ? " task" : " tasks") + " in the list\n");
+        horizontalLine();
+    }
+
+    private static void markUnmarkCommand(ArrayList<Task> tasks, String[] words) throws DukeException {
+        horizontalLine();
+        int taskNumber = Integer.parseInt(words[1]);
+        if (taskNumber > tasks.size()) {
+            throw new DukeException(ErrorMessage.INVALID_TASK.toString());
+        }
+
+        if (words[0].equals("mark")) {
+            tasks.get(taskNumber - 1).markDone();
+            System.out.println("Nice! I've marked this task as done:");
+        } else {
+            tasks.get(taskNumber - 1).markUndone();
+            System.out.println("OK, I've marked this task as not done yet:");
+        }
+        System.out.println(tasks.get(taskNumber - 1));
+        horizontalLine();
+    }
+
+    private static void writeToFile(ArrayList<Task> tasks) {
+        try {
+            FileOutputStream writeData = new FileOutputStream("duke.txt");
+            ObjectOutputStream writeStream = new ObjectOutputStream(writeData);
+
+            writeStream.writeObject(tasks);
+            writeStream.flush();
+            writeStream.close();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    private static ArrayList<Task> readFromFile(ArrayList<Task> tasks) {
+        try {
+            try {
+                FileInputStream readData = new FileInputStream("duke.txt");
+                ObjectInputStream readStream = new ObjectInputStream(readData);
+                @SuppressWarnings("unchecked")
+                ArrayList<Task> tasks2 = (ArrayList<Task>) readStream.readObject();
+                tasks = tasks2;
+                readStream.close();
+            } catch (FileNotFoundException error) {
+                System.out.println("File");
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        return tasks;
+    }
+
+    public static void main(String[] args) {
+        logoWithHello();
+        ArrayList<Task> tasks = new ArrayList<>();
+        File file = new File("duke.txt");
+        if (file.exists()) {
+            tasks = readFromFile(tasks);
+        }
+
+        greet();
+        Scanner in = new Scanner(System.in);
+        while (true) {
+            String userInput = in.nextLine();
+            if (userInput.equals("bye")) {
+                break;
+            }
+
+            try {
+                processCommands(tasks, userInput);
+            } catch (Exception error) {
+                horizontalLine();
+                System.out.println("Error message: " + error.getMessage());
+                horizontalLine();
+            }
+        }
+        writeToFile(tasks);
+        exit();
+    }
```

Comment:
The `Main` class is particularly dense. Please extract out classes and simplify the code